deterministic-bloom/src/lib.rs

```diff
@@ -2,29 +2,361 @@
 #![warn(missing_debug_implementations, missing_docs, rust_2018_idioms)]
 #![deny(unreachable_pub, private_in_public)]
 
-//! deterministic-bloom
+//! Deterministic Bloom filters
+//!
+//! This Crate is intented as a solid basis for cache reproducability
+//! and for underlying certain cryptographic primitives.
 
-/// Test utilities.
-#[cfg(any(test, feature = "test_utils"))]
-#[cfg_attr(docsrs, doc(cfg(feature = "test_utils")))]
+pub mod utils;
+
+#[cfg(feature = "test_utils")]
 pub mod test_utils;
 
-/// Add two integers together.
-pub fn add(a: i32, b: i32) -> i32 {
-    a + b
+use crate::utils::ByteArrayVisitor;
+use bitvec::prelude::BitArray;
+use miette::Diagnostic;
+use serde::{Deserialize, Serialize};
+use std::{fmt::Debug, ops::Index};
+use thiserror::Error;
+use xxhash_rust::xxh3;
+
+//------------------------------------------------------------------------------
+// Type Definitions
+//------------------------------------------------------------------------------
+
+/// The bloom filter is a probabilistic data structure that can be used to store a set of hashes.
+///
+/// `N` is the size of the bloom filter in bytes.
+///
+/// `K` is the number of bits to be set with each insert operation.
+///
+/// # Examples
+///
+/// ```
+/// use deterministic_bloom::BloomFilter;
+///
+/// let mut filter = BloomFilter::<256, 30>::default();
+/// filter.insert(&[0xF5u8; 32]);
+///
+/// assert!(filter.contains(&[0xF5u8; 32]));
+/// ```
+#[derive(Clone, PartialEq, Eq, PartialOrd)]
+pub struct BloomFilter<const N: usize, const K: usize> {
+    /// The underlying `BitArray`
+    pub bits: BitArray<[u8; N]>,
+}
+
+/// An iterator that generates indices into some bloom filter based on deterministic hashing of specified item.
+///
+/// `N` is the number of bytes in the bloom filter.
+/// This is used to restrict generated value within bloomfilter index space bounds.
+///
+/// # Examples
+///
+/// ```
+/// use deterministic_bloom::BloomFilter;
+///
+/// let filter = BloomFilter::<256, 30>::default();
+/// let indices = filter.hash_indices(&[0xF5u8; 32]);
+/// let indices = indices.collect::<Vec<_>>();
+///
+/// assert_eq!(indices.len(), 30);
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct HashIndexIterator<'a, T: AsRef<[u8]>, const N: usize> {
+    item: &'a T,
+    index: u64,
+}
+
+/// Errors for [BloomFilter] operations.
+#[derive(Error, Debug, Diagnostic)]
```

Comment:
minor: I'd specify this as thiserror::Error (instead of the import above), just to be explicit in the derive case.